### ISSUE TITLE
Add M2-mode to MELPA

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1097]
[MELPA](https://melpa.org/) is a package repository for Emacs package recipes. Each [recipe](https://github.com/melpa/melpa#recipe-format) gives the instructions for getting the package from the original source. For instance GAP has the following recipe for gap-mode:
```
(gap-mode
 :fetcher gitlab
 :repo "gvol/gap-mode"
 :files (:defaults "emacs.gaprc"))
```
This recipe points to the package hosted [here](https://bitbucket.org/gvol/gap-mode/src/default/).

The purpose of this issue is to propose moving the emacs directory to a separate M2-mode repository and add a recipe to the [MELPA repository](https://github.com/melpa/melpa). This would allow users to install M2-mode without installing Macaulay2 itself.

## Why?
The main advantage of doing this is that users could install M2-mode, then use `C-u F12` to choose how to run M2, for instance:
  - Remotely via SSH, e.g. `ssh math.umn.edu M2 --no-readline --print-width 125`
  - Remotely via SSH to a containerized version of Habanero!
  - Locally via Docker, e.g. `docker run -it --entrypoint M2 mahrud/macaulay2:v1.15`

Or for developers:
  - Locally from a build directory, without needing permission to install the artifacts on the system

This would mean that any machine running Emacs can run M2 via SSH or docker, and your files would still be saved locally!

The main thing that would need to be done is changing the build system to update the package on a separate M2-mode repository. This could be done via submodules, or we could manually update the M2-mode repository every release.

**Comment by DanGrayson** [comment]
Good idea!  But probably we should submit it first to ELPA, at https://elpa.gnu.org/packages/ , since then users wouldn't have to even add MELPA to the list of package archives.

**Comment by mahrud** [comment]
I considered that, but there are two big disadvantages: https://emacs.stackexchange.com/a/276
1. ELPA is strict about ownership of the package
2. ELPA does not support the same recipe system, so we would have to manually upload a new package every time

**Comment by DanGrayson** [comment]
1.  We can simply not copyright the files, see http://git.savannah.gnu.org/cgit/emacs/elpa.git/plain/README

2. How does the recipe system work?  (Fetching from the master branch every night seems not so good.)

**Comment by mahrud** [comment]
The update process for ELPA seems to be slower and more cumbersome. It is just more work, which is why there are 4.5k packages on MELPA versus 260 on ELPA. Given that Macaulay2 is neither a part of Emacs nor GNU, I really don't see an advantage to ELPA.

I don't understand the problem with fetching. The recipe model seems similar to homebrew and other distributed package managers. They query for changes from the specified repository a few times a day. We just update our `M2-mode` repository whenever necessary and that's it. This can be made automatic with a release script. Here are the details of the recipe system: https://github.com/melpa/melpa#recipe-format

By the way, adding MELPA to emacs is quite easy. You just add this to your `.emacs`:
```
(package-initialize)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")))
```

**Comment by DanGrayson** [comment]
Is there really a need to have an M2-mode repository, when the recipe could fetch the files from the M2 repository?

**Comment by mahrud** [comment]
We don't store the generated M2-mode files in the M2 repository. We would have to generate them and commit them somewhere. I figured a separate M2-mode repository embedded in the M2 repository as a git submodule and updated at every release would make things simpler because this way if at some point we decided to move the emacs folder in the M2 repository, or add or remove files, we wouldn't need to submit another PR to MELPA to change the recipe.

**Comment by DanGrayson** [comment]
Oh, right, some of those files are generated.  The entire set is present after a build in the subdirectory `usr-dist/common/share/emacs/site-lisp` of the top level build directory, so that would be a natural source for updating the releases.

But maybe it would be also good to just always store current versions of all the emacs files in the source tree, in `M2/Macaulay2/emacs/`.  One advantage of that is that developers would have them handy even before M2 gets built.

**Comment by DanGrayson** [comment]
By the way, it's already the case that some of the files in our source repository are generated: these are the example output files for packages whose examples require extra resources to run.

**Comment by mahrud** [comment]
I can see the merits of this. Is there any specific reason against having a separate repository?

Part of the reason why I prefer a separate repository is to keep things uniform between different editors. In particular, I've created this [Atom package](https://github.com/mahrud/atom-macaulay2/) which also has the "send to M2" feature. In order to register this package, it needs to be its own repository. Similar to the [language grammar](https://github.com/Macaulay2/language-macaulay2) repository which needs to be its own repository in order to work with [Github Linguist](https://github.com/github/linguist). This lets Github recognize .m2 files for the purpose of syntax highlighting and accurately showing what percentage of the repository is in M2 (currently Github says 56% C++, but after [this PR](https://github.com/github/linguist/pull/4844) is merged it'll be 84% M2, 10% C++!)

I was going to create a system to generate all of these (emacs, atom, vim, grammar, etc.) and push to the respective repositories, for instance by embedding them in the M2/submodules directory.

If we're going to keep it in the same repository, can I propose that we change the directory structure like so:
- contrib (where related, but not necessarily "Macaulay2" code is stored)
  - emacs
  - atom
  - link/html validator
  - etc.

Or:
- editors
  - emacs
  - atom
  - vim
  - etc.

And some of these would be submodues, perhaps emacs doesn't need to be.

**Comment by DanGrayson** [comment]
> I can see the merits of this. Is there any specific reason against having a separate repository?
> 

Yes, it's an additional maintenance task.

> Part of the reason why I prefer a separate repository is to keep things uniform between different editors. 

Having 'emacs' as a submodule, as you suggest, might not be so bad, for then building M2 would automatically modify the files in the submodule (according to what the latest list of M2 identifiers is), and all we'd have to do is remember to commit and push those changes periodically.  The command "git status" will tell us when things change in a submodule, so that won't be a problem.

Is that what you had in mind?

Can you implement this for both 'make' and 'cmake'?

Or what about leaving 'emacs' as it is and starting with submodules for 'atom', 'vim', etc?

**Comment by mahrud** [comment]
> Having 'emacs' as a submodule, as you suggest, might not be so bad, for then building M2 would automatically modify the files in the submodule (according to what the latest list of M2 identifiers is), and all we'd have to do is remember to commit and push those changes periodically. The command "git status" will tell us when things change in a submodule, so that won't be a problem.
> 
> Is that what you had in mind?

Yes, could you create a repository names `M2-emacs` and give me push access to it? I'll take care of adding a submodule in this repository.

> Can you implement this for both 'make' and 'cmake'?

Implement what exactly? Just add a command to fetch the submodule, or do you have something else in mind?

**Comment by DanGrayson** [comment]
Okay, I added the repository.  Just forget about the autotools-build for now.

**Comment by mahrud** [comment]
Thanks. I made the necessary changes in #1255 and opened https://github.com/melpa/melpa/pull/6984 to add M2-mode to MELPA.
